i2c/bmp280_i2c/bmp280_i2c.c

```diff
@@ -0,0 +1,260 @@
+  i2c_read_blocking(i2c_default, (addr & BMP280_READ_MODE), buf, 6,
+                    false);  // false - finished with bus
+
+  // store the 20 bit read in a 32 bit signed integer for conversion
+  *pressure = (buf[0] << 12) | (buf[1] << 4) | (buf[2] >> 4);
+  *temp = (buf[3] << 12) | (buf[4] << 4) | (buf[5] >> 4);
+}
+
+void bmp280_reset() {
+  // reset the device with the power-on-reset procedure
+  uint8_t buf[2] = {REG_RESET, 0xB6};
+  i2c_write_blocking(i2c_default, (addr & BMP280_WRITE_MODE), buf, 2, false);
+}
+
+float bmp280_convert_temp(int32_t *temp, struct bmp280_calib_param *params) {
+  // use the 32-bit floating point compensation implementation given in the
+  // datasheet
+  float var1, var2;
+  var1 = (((float)*temp) / 16384.0 - ((float)params->dig_t1) / 1024.0) *
+         ((float)params->dig_t2);
+  var2 = ((((float)*temp) / 131072.0 - ((float)params->dig_t1) / 8192.0) *
+          (((float)*temp) / 131072.0 - ((float)params->dig_t1) / 8192.0)) *
+         ((float)params->dig_t3);
+  params->t_fine = (int32_t)(var1 + var2);
+  return (var1 + var2) / 5120.0;
+}
+
+float bmp280_convert_pressure(int32_t *pressure,
+                              struct bmp280_calib_param *params) {
+  float var1, var2;
+  float converted = 0.0;
+
+  var1 = ((float)params->t_fine / 2.0) - 64000.0;
+  var2 = var1 * var1 * ((float)params->dig_p6) / 32768.0;
+  var2 = var2 + var1 * ((float)params->dig_p5) * 2.0;
+  var2 = (var2 / 4.0) + (((float)params->dig_p4) * 65536.0);
+  var1 = (((float)params->dig_p3) * var1 * var1 / 524288.0 +
+          ((float)params->dig_p2) * var1) /
+         524288.0;
+  var1 = (1.0 + var1 / 32768.0) * ((float)params->dig_p1);
+
+  // avoid division by 0
+  if (var1 < 0 || var1 > 0) {
+    converted = 1048576.0 - (float)*pressure;
+    converted = (converted - (var2 / 4096.0)) * 6250.0 / var1;
+    var1 = ((float)params->dig_p9) * converted * converted / 2147483648.0;
+    var2 = converted * ((float)params->dig_p8) / 32768.0;
+
+    converted += (var1 + var2 + ((float)params->dig_p7)) / 16.0;
+    // results in kPa
+    return converted / 1000.0;
+  } else {
+    return 0.0;
+  }
+}
+
+void bmp280_get_calib_params(struct bmp280_calib_param *params) {
+  // raw temp and pressure values need to be calibrated according to
+  // parameters generated during the manufacturing of the sensor
+  // there are 3 temperature params, and 9 pressure params, each with a LSB
+  // and MSB register, so we read from 24 registers
+
+  uint8_t buf[NUM_CALIB_PARAMS] = {0};
+  i2c_write_blocking(i2c_default, (addr & BMP280_WRITE_MODE), &REG_DIG_T1_LSB,
+                     1, true);  // true to keep master control of bus
+  // read in one go as register addresses auto-increment
+  i2c_read_blocking(i2c_default, (addr & BMP280_READ_MODE), buf,
+                    NUM_CALIB_PARAMS, false);  // false, we're done reading
+
+  // store these in a struct for later use
+  params->dig_t1 = (uint16_t)(buf[1] << 8) | buf[0];
```

Comment:
I'm not sure there is any particular benefit to using individual fields rather than an array, and then you could have a loop. (the unsigned/signed could be handled in the conversion functions which are _magic_ anyway

Reply:
Note this would let you use a loop which is going to be much smaller code wise... of course feel free to plug the code into [compiler explorer](https://godbolt.org/) to see if gcc can/does optimize it back into a loop

Reply:
I chose to go for individual fields for clarity so the person reading the code could see exactly where each parameter was being stored. I think hiding all that in an array would be good code-wise but would seem a bit cryptic with magic indices (unless we used `#define`s for the indices), not sure.

Reply:
yeah i don't feel strongly; my usual self is very at odds with the way we want to write examples. I just wasn't sure how much more helpful these numbers were, and yes I was thinking #defines for array indexes perhaps, but then again I haven't read the data sheet and therefore perhaps these are all specifically heralded.

Reply:
Fair, of note that the BME280 SPI example does something similar, which I only found out after I wrote this example but interesting coincidence.